synapse/event_auth.py

```diff
@@ -342,16 +351,41 @@ def _is_membership_change_allowed(
         # * They are not banned.
         # * They are accepting a previously sent invitation.
         # * They are already joined (it's a NOOP).
-        # * The room is public or restricted.
+        # * The room is public.
+        # * The room is restricted and the user meets the allows rules.
         if event.user_id != target_user_id:
             raise AuthError(403, "Cannot force another user to join.")
         elif target_banned:
             raise AuthError(403, "You are banned from this room")
-        elif join_rule == JoinRules.PUBLIC or (
+        elif join_rule == JoinRules.PUBLIC:
+            pass
+        elif (
             room_version.msc3083_join_rules
             and join_rule == JoinRules.MSC3083_RESTRICTED
         ):
-            pass
+            # This is the same as public, but the event must be signed by a server
+            # whose users could issue invites.
+            #
+            # Signatures are only checked once the event is fully created, e.g.
+            # not during make_join,
+            #
+            # Note that if the caller is in the room or invited, then they do
+            # not need to meet the allow rules.
+            if do_sig_check and not caller_in_room and not caller_invited:
+                # Find the servers of any users who could issue invites.
+                authorised_servers = get_servers_from_users(
+                    get_users_which_can_issue_invite(auth_events)
+                )
+
+                # Ensure one of the signatures is from one of the authorised servers.
+                # Note that it was previously checked that the signatures are
+                # valid.
+                for signing_server in event.signatures:
+                    if signing_server in authorised_servers:
+                        break
```

Comment:
We need to add a check in `federation_base._check_sigs_on_pdu` for a valid sig, we don't automatically check all signatures on events.

Reply:
Do we only require a single valid signature?

Reply:
I think that means we *don't* need to check the signature here then, does that seem correct?

Reply:
Actually this check should morph into ensuring that the user provided has the power to invite, I think.

Reply:
Yup